File: api/scalars/year.py
from re import compile
from graphene.types import Scalar
from graphql.language import ast
from graphql import GraphQLError

from functions.error_messages import *

YEAR_REGEX = r"^\d{4}$"

YEAR_REGEX_CHECK = compile(YEAR_REGEX)


class Year(Scalar):
    """
    A field that conforms to a 4 digit integer
    """

    @staticmethod
    def serialize(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type("String", value))

        if not YEAR_REGEX_CHECK.search(value):
            raise GraphQLError(scalar_error_type("Year", value))

        return value

    @staticmethod
    def parse_value(value):
        if not isinstance(value, str):
            raise GraphQLError(scalar_error_type("String", value))

        if not YEAR_REGEX_CHECK.search(value):
            raise GraphQLError(scalar_error_type("Year", value))

        return value

    @staticmethod
    def parse_literal(node):
        if not isinstance(node, ast.StringValue):
            raise GraphQLError(
                scalar_error_only_types("strings", "Years", str(type(node)))
            )

        if not YEAR_REGEX_CHECK.search(node.value):
            raise GraphQLError(scalar_error_type("Year", node.value))

        return node.value

File: api-js/src/scalars/year.js
import { GraphQLScalarType, Kind, GraphQLError } from 'graphql'

const validate = (value) => {
  const YEAR_REGEX = /^\d{4}$/

  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${typeof value}`)
  }

  if (!YEAR_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid year: ${value}`)
  }
  return value
}

export const Year = new GraphQLScalarType({
  name: 'Year',
  description: 'A field that conforms to a 4 digit integer.',
  serialize: validate,
  parseValue: validate,

  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Can only validate strings as year but got a: ${ast.kind}`,
      )
    }
    return validate(ast.value)
  },
})
